icicle-cpu/src/pe.rs

```diff
@@ -0,0 +1,205 @@
+                            let old_val = u32::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add(relocation_offset as u32);
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;                            
+                        },
+                        IMAGE_REL_BASED_DIR64 => {
+                            let mut buf = vec![0_u8; 8];
+                            cpu.mem.read_bytes(addr, &mut buf[..], perm::READ)
+                                .map_err(|e| format!("Failed to read relocation data: {e:?}"))?;                            
+                            let old_val = u64::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add(relocation_offset);                            
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;
+                        },
+                        _ => {
+                            return Err(String::from(format!("Relocation type {:#06x} not supported", reloc.typ)));
+                        }
+                    }
+                }
+            }
+        }
+        else {
+            return Err(String::from("PE file does not contain relocation data, unable to relocate file"));
+        }
+    }
+    
+
+    // create result object
+    let binary = PeMetadata {
+        relocation_offset: relocation_offset,
+        entry_ptr: nt_headers.optional_header().address_of_entry_point() as u64 + base_addr,
+        length: layout.size,
+        base_ptr: base_addr
+    };
+    
+    let mut debug_info = DebugInfo::default();
+    debug_info.add_file(data, relocation_offset)?;
+    Ok(LoadedPe { binary, debug_info })
+}
+
+// Retrives the base address and layout requirements of the PE when it is loaded into memory.
+fn get_layout<H>(nt_headers: &H) -> (u64, AllocLayout)
+where
+    H: object::read::pe::ImageNtHeaders,
+{
+    let optional_header = nt_headers.optional_header();
+    let image_base = optional_header.image_base();
+    let size = optional_header.size_of_image() as u64;
+    let align = optional_header.section_alignment() as u64;
+    (image_base, AllocLayout { addr: Some(image_base), size, align })
+}
+
+fn get_permission(section: &ImageSectionHeader) -> u8 {
+    let flags: u32 = section.characteristics.get(object::LittleEndian).into();    
+    let mut perm = perm::MAP;
+    perm |= if (flags & pe::IMAGE_SCN_MEM_READ) == 0 { perm::NONE } else { perm::READ };
+    perm |= if (flags & pe::IMAGE_SCN_MEM_WRITE) == 0 { perm::NONE } else { perm::WRITE };
+    perm |= if (flags & pe::IMAGE_SCN_MEM_READ) == 0 { perm::NONE } else { perm::EXEC };
```

Comment:
Should this be:

```suggestion
    perm |= if (flags & pe::IMAGE_SCN_MEM_EXECUTE) == 0 { perm::NONE } else { perm::EXEC };
```

Instead?

Reply:
good catch!